python/dgl/dataloading/dataloader.py

```diff
@@ -485,23 +487,92 @@ def create_tensorized_dataset(indices, batch_size, drop_last, use_ddp, ddp_seed)
         return TensorizedDataset(indices, batch_size, drop_last)
 
 
+def _get_device(device):
+    device = torch.device(device)
+    if device.type == 'cuda' and device.index is None:
+        device = torch.device('cuda', torch.cuda.current_device())
+    return device
+
+
 class DataLoader(torch.utils.data.DataLoader):
     """DataLoader class."""
     def __init__(self, graph, indices, graph_sampler, device='cpu', use_ddp=False,
                  ddp_seed=0, batch_size=1, drop_last=False, shuffle=False,
-                 use_prefetch_thread=False, use_alternate_streams=True, **kwargs):
+                 use_prefetch_thread=None, use_alternate_streams=None,
+                 pin_prefetcher=None, use_uva=False, **kwargs):
+        # (BarclayII) I hoped that pin_prefetcher can be merged into PyTorch's native
+        # pin_memory argument.  But our neighbor samplers and subgraph samplers
+        # return indices, which could be CUDA tensors (e.g. during UVA sampling)
+        # hence cannot be pinned.  PyTorch's native pin memory thread does not ignore
+        # CUDA tensors when pinning and will crash.  To enable pin memory for prefetching
+        # features and disable pin memory for sampler's return value, I had to use
+        # a different argument.  Of course I could change the meaning of pin_memory
+        # to pinning prefetched features and disable pin memory for sampler's returns
+        # no matter what, but I doubt if it's reasonable.
         self.graph = graph
+        num_workers = kwargs.get('num_workers', 0)
 
         try:
             if isinstance(indices, Mapping):
                 indices = {k: (torch.tensor(v) if not torch.is_tensor(v) else v)
                            for k, v in indices.items()}
+                indices_device = next(iter(indices.values())).device
             else:
                 indices = torch.tensor(indices) if not torch.is_tensor(indices) else indices
+                indices_device = indices.device
         except:     # pylint: disable=bare-except
             # ignore when it fails to convert to torch Tensors.
             pass
 
+        self.device = _get_device(device)
+
+        # Check graph and indices device as well as num_workers
+        if use_uva:
+            if self.graph.device.type != 'cpu':
+                raise ValueError('Graph must be on CPU if UVM sampling is enabled.')
+            if num_workers > 0:
+                raise ValueError('num_workers must be 0 if UVM sampling is enabled.')
+
+            self.graph.create_formats_()
+            self.graph.pin_memory_()
+            indices = recursive_apply(indices, lambda x: x.to(self.device))
+        else:
+            if self.graph.device != indices_device:
```

Comment:
Are we still trying to support GraphStorage-like objects? If so, for situations where the graph-structure is a different device than the `Frame`s, what should this value return? The alternate streams and prefetcher arguments are specific to the features, so if I have the graph structure in shared memory on the CPU, and features spread across the GPUs (or vice versa), this logic is problematic.

Reply:
Changes to only perform sanity checks on DGLGraphs.